### PR TITLE
feat: migration of categories - s3,dynamo,lambda,appsync

### PR DESCRIPTION
**Pull request by kaustavghosh06**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-api/provider-utils/awscloudformation/index.js [hunk 1]
```diff
@@ -151,4 +151,19 @@ async function updateResource(context, category, service) {
...
 
-module.exports = { addResource, updateResource, console };
+async function migrateResource(context, projectPath, service, resourceName) {
+  serviceMetadata = JSON.parse(fs.readFileSync(`${__dirname}/../supported-services.json`))[service];
+  const { serviceWalkthroughFilename } = serviceMetadata;
+  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
+  const { migrate } = require(serviceWalkthroughSrc);
+
+  if (!migrate) {
+    console.log(`No migration required for ${resourceName}`);
```
Comment (UnleashedMind) [inline packages/amplify-category-api/provider-utils/awscloudformation/index.js]: context is in scope, use context.print.info

File: packages/amplify-category-api/provider-utils/awscloudformation/index.js [hunk 2]
```diff
@@ -151,4 +151,19 @@ async function updateResource(context, category, service) {
...
+  }
+
+  return await migrate(context, projectPath, resourceName);
```
Comment (UnleashedMind) [inline packages/amplify-category-api/provider-utils/awscloudformation/index.js]: if `migrate` can be undefined or null, then this line needs to be combined with the previous if block, call `migrate` only if it's defined.

File: packages/amplify-category-function/index.js [hunk 3]
```diff
@@ -18,7 +18,36 @@ async function console(context) {
...
+  const { projectPath, amplifyMeta } = context.migrationInfo;
+  const migrateResourcePromises = [];
+  Object.keys(amplifyMeta).forEach((categoryName) => {
+    if (categoryName === category) {
+      Object.keys(amplifyMeta[category]).forEach((resourceName) => {
+        try {
+          const providerController = require(`./provider-utils/${amplifyMeta[category][resourceName].providerPlugin}/index`);
+          if (providerController) {
+            migrateResourcePromises.push(providerController.migrateResource(
+              projectPath,
```
Comment (UnleashedMind) [inline packages/amplify-category-function/index.js]: add context into the function parameter to be consistent, so the method does not need to use console.log(.), along with other usages of the context.

File: packages/amplify-category-function/provider-utils/awscloudformation/index.js [hunk 4]
```diff
@@ -199,4 +199,18 @@ async function invoke(context, category, service, resourceName) {
...
 
-module.exports = { addResource, invoke };
+function migrateResource(projectPath, service, resourceName) {
+  serviceMetadata = JSON.parse(fs.readFileSync(`${__dirname}/../supported-services.json`))[service];
+  const { serviceWalkthroughFilename } = serviceMetadata;
+  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
+  const { migrate } = require(serviceWalkthroughSrc);
+
+  if (!migrate) {
+    console.log(`No migration required for ${resourceName}`);
```
Comment (UnleashedMind) [inline packages/amplify-category-function/provider-utils/awscloudformation/index.js]: use context.print.info

File: packages/amplify-category-function/provider-utils/awscloudformation/index.js [hunk 5]
```diff
@@ -199,4 +199,18 @@ async function invoke(context, category, service, resourceName) {
...
+  }
+
+  return migrate(projectPath, resourceName);
```
Comment (UnleashedMind) [inline packages/amplify-category-function/provider-utils/awscloudformation/index.js]: if migrate can be undefined or null, then this line needs to be combined with the previous if block, call migrate only if it's defined.

File: packages/amplify-category-storage/provider-utils/awscloudformation/index.js [hunk 6]
```diff
@@ -31,5 +31,18 @@ function updateResource(context, category, service) {
...
+  const serviceMetadata = JSON.parse(fs.readFileSync(`${__dirname}/../supported-services.json`))[service];
+  const { serviceWalkthroughFilename } = serviceMetadata;
+  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
+  const { migrate } = require(serviceWalkthroughSrc);
+
+  if (!migrate) {
+    console.log(`No migration required for ${resourceName}`);
+  }
+
+  return migrate(projectPath, resourceName);
```
Comment (UnleashedMind) [inline packages/amplify-category-storage/provider-utils/awscloudformation/index.js]: if migrate can be undefined or null, then this line needs to be combined with the previous if block, call migrate only if it's defined.

File: packages/amplify-category-storage/provider-utils/awscloudformation/index.js [hunk 7]
```diff
@@ -31,5 +31,18 @@ function updateResource(context, category, service) {
...
 }
 
+function migrateResource(projectPath, service, resourceName) {
+  const serviceMetadata = JSON.parse(fs.readFileSync(`${__dirname}/../supported-services.json`))[service];
+  const { serviceWalkthroughFilename } = serviceMetadata;
+  const serviceWalkthroughSrc = `${__dirname}/service-walkthroughs/${serviceWalkthroughFilename}`;
+  const { migrate } = require(serviceWalkthroughSrc);
+
+  if (!migrate) {
+    console.log(`No migration required for ${resourceName}`);
```
Comment (UnleashedMind) [inline packages/amplify-category-storage/provider-utils/awscloudformation/index.js]: use context.print.info to be consistent

File: packages/amplify-category-api/provider-utils/awscloudformation/index.js (hunk 1 above)
Comment (kaustavghosh06) [reply packages/amplify-category-api/provider-utils/awscloudformation/index.js]: Fixed.

File: packages/amplify-category-api/provider-utils/awscloudformation/index.js (hunk 2 above)
Comment (kaustavghosh06) [reply packages/amplify-category-api/provider-utils/awscloudformation/index.js]: Fixed.

File: packages/amplify-category-function/index.js (hunk 3 above)
Comment (kaustavghosh06) [reply packages/amplify-category-function/index.js]: Fixed.

File: packages/amplify-category-function/provider-utils/awscloudformation/index.js (hunk 4 above)
Comment (kaustavghosh06) [reply packages/amplify-category-function/provider-utils/awscloudformation/index.js]: Fixed.

**Pull request merged**